# Backpropagation

The neural network looks like this:

![title](img/NN_structure.png)

## Forwardpass

To begin, lets see what the neural network currently predicts given the weights and biases above and inputs of 0.05 and 0.10. To do this we’ll feed those inputs forward though the network.

\begin{equation*}
net_{h1} = w_1 * i_1 + w_2 * i_2 + b_1 * 1\\\
net_{h2} = w_3 * i_1 + w_4 * i_2 + b_1 * 1
\end{equation*}

In [1]:
i1 = 0.05
i2 = 0.10

b1 = 0.35

w1 = 0.15
w2 = 0.20
w3 = 0.25
w4 = 0.30

o1 = 0.01
o2 = 0.99

b2 = 0.60

w5 = 0.4
w6 = 0.45
w7 = 0.5
w8 = 0.55

In [2]:
net_h1 = w1 * i1 + w2 * i2 + b1 * 1
net_h2 = w3 * i1 + w4 * i2 + b1 * 1

In [3]:
net_h1, net_h2

(0.3775, 0.39249999999999996)

---
We then squash it using the logistic function to get the output of h_1:

\begin{equation*}
out_{h1} = \frac{1}{1 + e^{-net_{h1}}} \\\
\\\
out_{h1} = \frac{1}{1 + e^{-0.3775}} \\\
\end{equation*}

We do the same for h_2:

---

In [4]:
from math import e

def logistic_function(x):
    
    return 1 / (1 + e ** (-x))

In [5]:
out_h1 = logistic_function(net_h1)
out_h2 = logistic_function(net_h2)

In [6]:
out_h1, out_h2

(0.5932699921071872, 0.596884378259767)

---
We repeat this process for the output layer neurons, using the output from the hidden layer neurons as inputs.

---

\begin{equation*}
net_{o1} = w_5 * out_{h1} + w_6 * out_{h2} + b_2 * 1\\\
net_{o2} = w_7 * out_{h1} + w_8 * out_{h2} + b_2 * 1
\end{equation*}

---

In [7]:
net_o1 = w5 * out_h1 + w6 * out_h2 + b2 * 1
net_o2 = w7 * out_h1 + w8 * out_h2 + b2 * 1

In [8]:
net_o1, net_o2

(1.10590596705977, 1.2249214040964653)

---
We then squash it using the logistic function to get the output of o_1:

\begin{equation*}
out_{o1} = \frac{1}{1 + e^{-net_{o1}}} \\\
\\\
out_{h1} = \frac{1}{1 + e^{-1.10590}} \\\
\end{equation*}

We do the same for o_2:

---

In [9]:
out_o1 = logistic_function(net_o1)
out_o2 = logistic_function(net_o2)

In [10]:
out_o1, out_o2

(0.7513650695523157, 0.7729284653214625)

---
Forwardpropagation finishes here.

---

## Calculating the Total Error

---
We can now calculate the error for each output neuron using the squared error function and sum them to get the total error:

---

\begin{equation*}
\sum_{}^{} {\frac{1}{2}(target_{o1} - out_{o1})^2}
\end{equation*}

---
For example, the target output for o_1 is 0.01 but the neural network output 0.75136507, therefore its error is:

\begin{equation*}
E_{o1} = \sum_{}^{} {\frac{1}{2}(0.01 - 0.75136507)^2} = 0.274811083
\end{equation*}

Repeating this process for o_2 (remembering that the target is 0.99) we get:

\begin{equation*}
E_{o2} = 0.023560026
\end{equation*}

The total error for the neural network is the sum of these errors:

\begin{equation*}
E_{total} = E_{o1} + E_{o2}
\end{equation*}


In [11]:
E_o1 = 0.5 * (o1 - out_o1) ** 2
E_o2 = 0.5 * (o2 - out_o2) ** 2

In [12]:
E_o1, E_o2

(0.274811083176155, 0.023560025583847746)

In [13]:
E_total = E_o1 + E_o2

In [14]:
E_total

0.2983711087600027

## The Backward Pass

Our goal with backpropagation is to update each of the weights in the network so that they cause the actual output to be closer the target output, thereby minimizing the error for each output neuron and the network as a whole.

### Output Layer

Consider $w_5$. We want to know how much a change in $w_5$ affects the total error, aka $\frac{\partial E_{total}}{\partial w_{5}}$.

By applying the chain rule we know that:

\begin{equation*}
\frac{\partial E_{total}}{\partial w_{5}} = \frac{\partial E_{total}}{\partial out_{o1}} * \frac{\partial out_{o1}}{\partial net_{o1}} * \frac{\partial net_{o1}}{\partial w_{5}}
\end{equation*}

Visually, here’s what we’re doing:

![title](img/backwardpass.png)

We need to figure out each piece in this equation.

<ol>
<li>
<b>First, how much does the total error change with respect to the output?</b>

\begin{equation*}
E_{total} = \frac{1}{2}(target_{o1} - out_{o1})^{2} + \frac{1}{2}(target_{o2} - out_{o2})^{2} \\\
\\\
\frac{\partial E_{total}}{\partial out_{o1}} = 2 * \frac{1}{2}(target_{o1} - out_{o1})^{2 - 1} * -1 + 0 \\\
\\\
\frac{\partial E_{total}}{\partial out_{o1}} = -(target_{o1} - out_{o1}) = -(0.01 - 0.75136507) = 0.74136507 \\\
\end{equation*}

<br>

---

</li>
<li>
<b>Next, how much does the output of $o_1$ change with respect to its total net input?</b>

The partial derivative of the logistic function is the output multiplied by 1 minus the output:

\begin{equation*}
out_{o1} = \frac{1}{1+e^{-net_{o1}}} \\\
\\\
\frac{\partial out_{o1}}{\partial net_{o1}} = out_{o1}(1 - out_{o1}) = 0.75136507(1 - 0.75136507) = 0.186815602 \\\
\end{equation*}

<br>

---

</li>
<li>
<b>Finally, how much does the total net input of $net_{o1}$ change with respect to $w_5$?</b>

\begin{equation*}
net_{o1} = w_5 * out_{h1} + w_6 * out_{h2} + b_2 * 1 \\\
\\\
\frac{\partial net_{o1}}{\partial w_{5}} = 1 * out_{h1} * w_5^{(1 - 1)} + 0 + 0 = out_{h1} = 0.593269992 \\\
\end{equation*}

<br>

---

</li>
</ol>
<b>Putting it all together:</b>

\begin{equation*}
\frac{\partial E_{total}}{\partial w_{5}} = \frac{\partial E_{total}}{\partial out_{o1}} * \frac{\partial out_{o1}}{\partial net_{o1}} * \frac{\partial net_{o1}}{\partial w_{5}}\\\
\\\
\frac{\partial E_{total}}{\partial w_{5}} = 0.74136507 * 0.186815602 * 0.593269992 = 0.082167041
\\\
\end{equation*}

<br>
You’ll often see this calculation combined in the form of the [delta rule](https://en.wikipedia.org/wiki/Delta_rule):

\begin{equation*}
\frac{\partial E_{total}}{\partial w_{5}} = -(target_{o1} - out_{o1}) * out_{o1}(1 - out_{o1}) * out_{h1}
\end{equation*}

<br>
Alternatively, we have $\frac{\partial E_{total}}{\partial out_{o1}}$ and $ \frac{\partial E_{total}}{\partial net_{o1}}$ which can be written as $\frac{\partial E_{total}}{\partial net_{o1}}$, aka $ \delta_{o1}$ (the Greek letter delta) aka the node delta. We can use this to rewrite the calculation above:

\begin{equation*}
\delta_{o1} = \frac{\partial E_{total}}{\partial out_{o1}} * \frac{\partial out_{o1}}{\partial net_{o1}} = \frac{\partial E_{total}}{\partial net_{o1}}
\end{equation*}

\begin{equation*}
\delta_{o1} = -(target_{o1} - out_{o1}) * out_{o1}(1 - out_{o1})
\end{equation*}

<br>
Therefore:

\begin{equation*}
\frac{\partial E_{total}}{\partial w_{5}} = \delta_{o1} out_{h1}
\end{equation*}

<br>
Some sources extract the negative sign from \delta so it would be written as:

\begin{equation*}
\frac{\partial E_{total}}{\partial w_{5}} = -\delta_{o1} out_{h1}
\end{equation*}

<br>
To decrease the error, we then subtract this value from the current weight (optionally multiplied by some learning rate, eta, which we’ll set to 0.5):

\begin{equation*}
w_5^{+} = w_5 - \eta * \frac{\partial E_{total}}{\partial w_{5}} = 0.4 - 0.5 * 0.082167041 = 0.35891648
\end{equation*}

<br>
We can repeat this process to get the new weights $w_6$, $ w_7$, and $w_8$:

$w_6^{+} = 0.408666186$<br>
$w_7^{+} = 0.511301270$<br>
$w_8^{+} = 0.561370121$<br>

We perform the actual updates in the neural network after we have the new weights leading into the hidden layer neurons (ie, we use the original weights, not the updated weights, when we continue the backpropagation algorithm below).

In [15]:
delta_o1 = -(o1 - out_o1) * out_o1 * (1 - out_o1)
delta_o2 = -(o2 - out_o2) * out_o2 * (1 - out_o2)

In [16]:
delta_o1, delta_o2

(0.13849856162855695, -0.038098236516556236)

In [17]:
w5_contrib = delta_o1 * out_h1
w6_contrib = delta_o1 * out_h2

w7_contrib = delta_o2 * out_h1
w8_contrib = delta_o2 * out_h2

b2_contrib = delta_o1 + delta_o2

In [18]:
w5_contrib, w6_contrib, w7_contrib, w8_contrib, b2_contrib

(0.08216704056423077,
 0.08266762784753325,
 -0.02260254047747507,
 -0.022740242215978222,
 0.10040032511200071)

In [19]:
alfa = 0.5

w5_updated = w5 - alfa * w5_contrib
w6_updated = w6 - alfa * w6_contrib
w7_updated = w7 - alfa * w7_contrib
w8_updated = w8 - alfa * w8_contrib

b2_updated = b2 - alfa * b2_contrib

In [20]:
w5_updated, w6_updated, w7_updated, w8_updated, b2_updated

(0.35891647971788465,
 0.4086661860762334,
 0.5113012702387375,
 0.5613701211079891,
 0.5497998374439996)

### Hidden Layer

Next, we’ll continue the backwards pass by calculating new values for $w_1$, $w_2$, $w_3$, and $w_4$.

Big picture, here’s what we need to figure out:

\begin{equation*}
\frac{\partial E_{total}}{\partial w_{1}} = \frac{\partial E_{total}}{\partial out_{h1}} * \frac{\partial out_{h1}}{\partial net_{h1}} * \frac{\partial net_{h1}}{\partial w_{1}}
\end{equation*}

Visually:

<br>
![title](img/hidden_layer_bp.png)
<br>

We’re going to use a similar process as we did for the output layer, but slightly different to account for the fact that the output of each hidden layer neuron contributes to the output (and therefore error) of multiple output neurons. We know that $out_{h1}$ affects both $out_{o1}$ and $out_{o2}$ therefore the $\frac{\partial E_{total}}{\partial out_{h1}}$ needs to take into consideration its effect on the both output neurons:

\begin{equation*}
\frac{\partial E_{total}}{\partial out_{h1}} = \frac{\partial E_{o1}}{\partial out_{h1}} + \frac{\partial E_{o2}}{\partial out_{h1}}
\end{equation*}

<br>
Starting with $\frac{\partial E_{o1}}{\partial out_{h1}}$:

\begin{equation*}
\frac{\partial E_{o1}}{\partial out_{h1}} = \frac{\partial E_{o1}}{\partial net_{o1}} * \frac{\partial net_{o1}}{\partial out_{h1}}
\end{equation*}

<br>
We can calculate $\frac{\partial E_{o1}}{\partial net_{o1}}$ using values we calculated earlier:

\begin{equation*}
\frac{\partial E_{o1}}{\partial net_{o1}} = \frac{\partial E_{o1}}{\partial out_{o1}} * \frac{\partial out_{o1}}{\partial net_{o1}} = 0.74136507 * 0.186815602 = 0.18498562
\end{equation*}

<br>
And $\frac{\partial net_{o1}}{\partial out_{h1}}$ is equal to $w_5$:

\begin{equation*}
net_{o1} = w_5 * out_{h1} + w_6 * out_{h2} + b_2 * 1 \\\
\\\
\frac{\partial net_{o1}}{\partial out_{h1}} = w_5 = 0.40 \\\
\end{equation*}

<br>
Plugging them in:

\begin{equation*}
\frac{\partial E_{o1}}{\partial out_{h1}} = \frac{\partial E_{o1}}{\partial net_{o1}} * \frac{\partial net_{o1}}{\partial out_{h1}} = 0.138498562 * 0.40 = 0.055399425
\end{equation*}

<br>
Following the same process for $\frac{\partial E_{o2}}{\partial out_{h1}}$, we get:

\begin{equation*}
\frac{\partial E_{o2}}{\partial out_{h1}} = -0.019049119
\end{equation*}

Therefore:

\begin{equation*}
\frac{\partial E_{total}}{\partial out_{h1}} = \frac{\partial E_{o1}}{\partial out_{h1}} + \frac{\partial E_{o2}}{\partial out_{h1}} = 0.055399425 + -0.019049119 = 0.036350306
\end{equation*}

<br>
Now that we have $\frac{\partial E_{total}}{\partial out_{h1}}$, we need to figure out $\frac{\partial out_{h1}}{\partial net_{h1}}$ and then $\frac{\partial net_{h1}}{\partial w}$ for each weight:

\begin{equation*}
out_{h1} = \frac{1}{1+e^{-net_{h1}}} \\\
\\\
\frac{\partial out_{h1}}{\partial net_{h1}} = out_{h1}(1 - out_{h1}) = 0.59326999(1 - 0.59326999 ) = 0.241300709 \\\
\end{equation*}

We calculate the partial derivative of the total net input to h_1 with respect to w_1 the same as we did for the output neuron:

\begin{equation*}
net_{h1} = w_1 * i_1 + w_3 * i_2 + b_1 * 1 \\\
\\\
\frac{\partial net_{h1}}{\partial w_1} = i_1 = 0.05 \\\
\end{equation*}


<br>
Putting it all together:

\begin{equation*}
\frac{\partial E_{total}}{\partial w_{1}} = \frac{\partial E_{total}}{\partial out_{h1}} * \frac{\partial out_{h1}}{\partial net_{h1}} * \frac{\partial net_{h1}}{\partial w_{1}} \\\
\\\
\frac{\partial E_{total}}{\partial w_{1}} = 0.036350306 * 0.241300709 * 0.05 = 0.000438568 \\\
\end{equation*}

<br>

***

You might also see this written as:

\begin{equation*}
\frac{\partial E_{total}}{\partial w_{1}} = (\sum\limits_{o}{\frac{\partial E_{total}}{\partial out_{o}} * \frac{\partial out_{o}}{\partial net_{o}} * \frac{\partial net_{o}}{\partial out_{h1}}}) * \frac{\partial out_{h1}}{\partial net_{h1}} * \frac{\partial net_{h1}}{\partial w_{1}} \\\
\\\
\frac{\partial E_{total}}{\partial w_{1}} = (\sum\limits_{o}{\delta_{o} * w_{ho}}) * out_{h1}(1 - out_{h1}) * i_{1} \\\
\\\
\frac{\partial E_{total}}{\partial w_{1}} = \delta_{h1}i_{1}\\\
\end{equation*}

---

We can now update $w_1$:

\begin{equation*}
w_1^{+} = w_1 - \eta * \frac{\partial E_{total}}{\partial w_{1}} = 0.15 - 0.5 * 0.000438568 = 0.149780716
\end{equation*}

<br>
Repeating this for $w_2$, $w_3$, and $w_4$:

\begin{equation*}
w_2^{+} = 0.19956143 \\\
w_3^{+} = 0.24975114 \\\
w_4^{+} = 0.29950229 \\\
\end{equation*}

<br>
Finally, we’ve updated all of our weights! When we fed forward the 0.05 and 0.1 inputs originally, the error on the network was 0.298371109. After this first round of backpropagation, the total error is now down to 0.291027924. It might not seem like much, but after repeating this process 10,000 times, for example, the error plummets to 0.0000351085. At this point, when we feed forward 0.05 and 0.1, the two outputs neurons generate 0.015912196 (vs 0.01 target) and 0.984065734 (vs 0.99 target).

In [21]:
# dEtotal/dout_o1 = dE_o1/dout_o1 + dE_o2/dout_o2
# tackling dE_o1/dout_o1 part:

dEo1_douto1 = -1 * (o1 - out_o1)

#tackling dE_o2/dout_o2 part:

dEo2_douto2 = -1 * (o2 - out_o2)

# adding Eo1 and Eo2 to find Etotal

Etotal = dEo1_douto1 + dEo2_douto2

In [22]:
# dout_o1/dnet_o1

douto1_dneto1 = out_o1 * (1 - out_o1)
douto2_dneto2 = out_o2 * (1 - out_o2)

In [23]:
# dnet_o1/dout_h1

dneto1_douth1 = w5
dneto2_douth1 = w7

dneto1_douth2 = w6
dneto2_douth2 = w8

In [24]:
# dout_h1/dnet_h1

douth1_dneth1 = out_h1 * (1 - out_h1)
douth2_dneth2 = out_h2 * (1 - out_h2)

In [25]:
#dnet_h1/dw_1

dneth1_dw1 = i1
dneth1_dw2 = i2

dneth2_dw3 = i1
dneth2_dw4 = i2

dneth1_db1 = 1
dneth2_db1 = 1

In [26]:
#putting them all together

w1_contrib = (
    (
        (dEo1_douto1 * douto1_dneto1 * dneto1_douth1) + 
        (dEo2_douto2 * douto2_dneto2 * dneto2_douth1)
    )
    * douth1_dneth1 * dneth1_dw1
)


w2_contrib = (
    (
        (dEo1_douto1 * douto1_dneto1 * dneto1_douth1) +
        (dEo2_douto2 * douto2_dneto2 * dneto2_douth1)
    ) 
    * douth1_dneth1 * dneth1_dw2
)


w3_contrib = (
    (
        (dEo1_douto1 * douto1_dneto1 * dneto1_douth2) + 
        (dEo2_douto2 * douto2_dneto2 * dneto2_douth2)
    ) 
    * douth2_dneth2 * dneth2_dw3
)


w4_contrib = (
    (
        (dEo1_douto1 * douto1_dneto1 * dneto1_douth2) + 
        (dEo2_douto2 * douto2_dneto2 * dneto2_douth2)
    ) 
    * douth2_dneth2 * dneth2_dw4
)


b1_contrib = (
    (
        delta_o1 * (
            (dneto1_douth1 * douth1_dneth1 * dneth1_db1) + 
            (dneto1_douth2 * douth2_dneth2 * dneth2_db1)
        ) +
        delta_o2 * (
            (dneto2_douth1 * douth1_dneth1 * dneth1_db1) +
            (dneto2_douth2 * douth2_dneth2 * dneth2_db1)
        )
    )
)

In [27]:
w1_contrib, w2_contrib, w3_contrib, w4_contrib, b1_contrib

(0.00043856773447434685,
 0.0008771354689486937,
 0.0004977127352608601,
 0.0009954254705217202,
 0.018725609394704128)

In [28]:
w1_updated = w1 - alfa * w1_contrib
w2_updated = w2 - alfa * w2_contrib
w3_updated = w3 - alfa * w3_contrib
w4_updated = w4 - alfa * w4_contrib

b1_updated = b1 - alfa * b1_contrib

In [29]:
w1_updated, w2_updated, w3_updated, w4_updated, b1_updated

(0.1497807161327628,
 0.19956143226552567,
 0.24975114363236958,
 0.29950228726473915,
 0.3406371953026479)

In [30]:
def epoch(w1,
          w2,
          w3,
          w4,
          b1,
          w5,
          w6,
          w7,
          w8,
          b2
):
    
    net_h1 = w1 * i1 + w2 * i2 + b1 * 1
    net_h2 = w3 * i1 + w4 * i2 + b1 * 1

    out_h1 = logistic_function(net_h1)
    out_h2 = logistic_function(net_h2)

    net_o1 = w5 * out_h1 + w6 * out_h2 + b2 * 1
    net_o2 = w7 * out_h1 + w8 * out_h2 + b2 * 1

    out_o1 = logistic_function(net_o1)
    out_o2 = logistic_function(net_o2)

    E_o1 = 0.5 * (o1 - out_o1) ** 2
    E_o2 = 0.5 * (o2 - out_o2) ** 2

    E_total = E_o1 + E_o2
    print(E_total)

    delta_o1 = -(o1 - out_o1) * out_o1 * (1 - out_o1)
    delta_o2 = -(o2 - out_o2) * out_o2 * (1 - out_o2)

    w5_contrib = delta_o1 * out_h1
    w6_contrib = delta_o1 * out_h2

    w7_contrib = delta_o2 * out_h1
    w8_contrib = delta_o2 * out_h2

    b2_contrib = delta_o1 + delta_o2

    alfa = 0.5

    w5_updated = w5 - alfa * w5_contrib
    w6_updated = w6 - alfa * w6_contrib
    w7_updated = w7 - alfa * w7_contrib
    w8_updated = w8 - alfa * w8_contrib

    b2_updated = b2 - alfa * b2_contrib

    # dEtotal/dout_o1 = dE_o1/dout_o1 + dE_o2/dout_o2
    # tackling dE_o1/dout_o1 part:

    dEo1_douto1 = -1 * (o1 - out_o1)

    #tackling dE_o2/dout_o2 part:

    dEo2_douto2 = -1 * (o2 - out_o2)

    # adding Eo1 and Eo2 to find Etotal

    Etotal = dEo1_douto1 + dEo2_douto2

    # dout_o1/dnet_o1

    douto1_dneto1 = out_o1 * (1 - out_o1)
    douto2_dneto2 = out_o2 * (1 - out_o2)

    # dnet_o1/dout_h1

    dneto1_douth1 = w5
    dneto2_douth1 = w7

    dneto1_douth2 = w6
    dneto2_douth2 = w8

    # dout_h1/dnet_h1

    douth1_dneth1 = out_h1 * (1 - out_h1)
    douth2_dneth2 = out_h2 * (1 - out_h2)

    #dnet_h1/dw_1

    dneth1_dw1 = i1
    dneth1_dw2 = i2

    dneth2_dw3 = i1
    dneth2_dw4 = i2

    dneth1_db1 = 1
    dneth2_db1 = 1

    #putting them all together

    w1_contrib = (
        (
            (dEo1_douto1 * douto1_dneto1 * dneto1_douth1) + 
            (dEo2_douto2 * douto2_dneto2 * dneto2_douth1)
        )
        * douth1_dneth1 * dneth1_dw1
    )


    w2_contrib = (
        (
            (dEo1_douto1 * douto1_dneto1 * dneto1_douth1) +
            (dEo2_douto2 * douto2_dneto2 * dneto2_douth1)
        ) 
        * douth1_dneth1 * dneth1_dw2
    )


    w3_contrib = (
        (
            (dEo1_douto1 * douto1_dneto1 * dneto1_douth2) + 
            (dEo2_douto2 * douto2_dneto2 * dneto2_douth2)
        ) 
        * douth2_dneth2 * dneth2_dw3
    )


    w4_contrib = (
        (
            (dEo1_douto1 * douto1_dneto1 * dneto1_douth2) + 
            (dEo2_douto2 * douto2_dneto2 * dneto2_douth2)
        ) 
        * douth2_dneth2 * dneth2_dw4
    )


    b1_contrib = (
        (
            delta_o1 * (
                (dneto1_douth1 * douth1_dneth1 * dneth1_db1) + 
                (dneto1_douth2 * douth2_dneth2 * dneth2_db1)
            ) +
            delta_o2 * (
                (dneto2_douth1 * douth1_dneth1 * dneth1_db1) +
                (dneto2_douth2 * douth2_dneth2 * dneth2_db1)
            )
        )
    )

    w1_contrib, w2_contrib, w3_contrib, w4_contrib, b1_contrib

    w1_updated = w1 - alfa * w1_contrib
    w2_updated = w2 - alfa * w2_contrib
    w3_updated = w3 - alfa * w3_contrib
    w4_updated = w4 - alfa * w4_contrib

    b1_updated = b1 - alfa * b1_contrib

    return (w1_updated, w2_updated, w3_updated, w4_updated, b1_updated,
            w5_updated, w6_updated, w7_updated, w8_updated, b2_updated)

In [31]:
for _ in range(20000):
    w1,w2,w3,w4,b1,w5,w6,w7,w8,b2 = epoch(w1,w2,w3,w4,b1,w5,w6,w7,w8,b2)

0.2983711087600027
0.2857508866992965
0.2727972641999933
0.25965950665650733
0.24650109802541337
0.23348846764026632
0.22077865301586658
0.20850799425100638
0.19678379368361937
0.1856801337796513
0.1752380107751284
0.16546900343098178
0.15636114950118693
0.14788562450079368
0.1400030923513818
0.13266902812313663
0.12583772807075122
0.11946502713801205
0.11350991807250185
0.10793533226893048
0.10270833822744369
0.09779997352081168
0.09318487460182723
0.08884081950787168
0.08474825773776462
0.08088987081921514
0.07725018551219504
0.07381524744646799
0.07057235438606374
0.06750984359676877
0.06461692565726448
0.06188355653335943
0.059300340148360386
0.05685845457497422
0.05454959605348296
0.05236593613135995
0.05030008822364901
0.048345080765248585
0.04649433485159678
0.04474164484764585
0.04308116090077957
0.04150737263996594
0.040015093600210994
0.03859944609636258
0.03725584639944027
0.03597999015538718
0.03476783804143728
0.03361560168787959
0.03251972990965193
0.03147689529803803
0.0

In [32]:
def predict(w1,w2,w3,w4,b1,w5,w6,w7,w8,b2):
    
    net_h1 = w1 * i1 + w2 * i2 + b1 * 1
    net_h2 = w3 * i1 + w4 * i2 + b1 * 1

    out_h1 = logistic_function(net_h1)
    out_h2 = logistic_function(net_h2)

    net_o1 = w5 * out_h1 + w6 * out_h2 + b2 * 1
    net_o2 = w7 * out_h1 + w8 * out_h2 + b2 * 1

    out_o1 = logistic_function(net_o1)
    out_o2 = logistic_function(net_o2)
    
    return out_o1, out_o2

In [33]:
out_o1, out_o2 = predict(w1,w2,w3,w4,b1,w5,w6,w7,w8,b2)
out_o1, out_o2

(0.010633774372091731, 0.9893773083483602)